File: RotateAndSearch.py
def bsearch(arr,low,high,key):
    mid = int((low+high)/2)
    if (low>high):
        return -1
    elif (low == high):
        return low
    elif (key == arr[mid]):
        return mid
    elif (key < arr[mid]):
        return bsearch(arr,low,mid-1,key)
    elif (key > arr[mid]):
        return bsearch(arr,mid+1,high,key)
def FindPivot(arr,low,high):
    mid = (low+high)/2
    if (low>high):
        return -1
    elif (low == high):
        return low
    elif (low == mid):
        return (mid)
    elif (mid < high and arr[mid] > arr[mid + 1]):
        
        print("Pivot found at",(mid))
        return (mid)
    elif (mid > low and arr[mid] < arr[mid-1]):
        print("Pivot found at",mid-1)
        return (mid-1)
    elif (arr[mid] > arr[low]):
        return FindPivot(arr,mid+1,high)
    return FindPivot(arr,low,mid-1)
def RotateAndSearch():
    count = 1
    ch = "y"
    flag = 1
    f = 0
    while(int(count) <= 200):
     
        
        
        if (ch == "y"):
            arr = [0]
            N = raw_input("Enter length of array:")
            g = raw_input("Enter element to be searched:")
            print("Element to be searched",g)
            
            for i in range(int(N)-1):
                arr.append(0)
            print("enter array elements one by one")
            for i in range(int(N)):
                arr[i] = input("enter element" + "" + str(i+1))
            print arr
            print ("lenth of array" + " " +str(len(arr)))
            d = FindPivot(arr,0,int(N)-1)
            if (d == -1):
                f = bsearch(arr,0,int(N)-1,g)
            elif (g > arr[0]):
                f = bsearch(arr,0,d,g)
            elif (g < arr[0]):
                f = bsearch(arr,d+1,int(N)-1,g)
            if (f == -1):
                print ("OOPS,NOT FOUND")
                flag = 0
                break
            else:
                print("Element found at",f)
        else:
            exit(0)
        print ("no. of test cases done:" + str(count))
        if (int(count) == 199):
            print "last test case remaining"
        count = count + 1
        ch = raw_input("Do you want to test more? y/n:")
        
        
RotateAndSearch()

File: RotateAndSearch.java
import java.util.Scanner;
import java.util.Arrays;
public class RotateAndSearch {
    public static void main(String args[]) {
        int[] arr1;
        int N, d;
        boolean flag = true;

        int count;

        Scanner scan = new Scanner(System.in);
        System.out.println("Enter number if test cases");
        count = scan.nextInt();
        while (count >= 0){
            System.out.println("Enter length of array");
            N = scan.nextInt();
            System.out.println("Enter Element to be searched");
            d = scan.nextInt();
            arr1 = new int[N];
            System.out.println("Enter elments one by one");
            for(int i = 0; i < N; i++){
                arr1[i] = scan.nextInt();
            }
            int pivot = FindPivot(arr1,0,N-1);
            if (pivot == -1){
                System.out.println("Pivot not found!");
                bsearch(arr1,0,N-1,d);
            }
            else if(d == arr1[pivot]){
                System.out.println("Element found at" + pivot);
            }
            else if(arr1[0] > d){
                bsearch(arr1,pivot+1,N-1,d);
            }
            else{
                bsearch(arr1,0,pivot-1,d);
            }
            count = count + 1;
        }
    }
    public static int FindPivot(int[] arr, int low, int high){
        int mid;
        mid = (low+high)/2;
        
        if(mid < high && arr[mid] > arr[mid+1]){
            return (mid+1);
        }
        else if(mid > low && arr[mid] < arr[mid-1]){
            return mid;

        }
        else if (low > high || low == high){
            return -1;
        }
        else if(arr[low] < arr[mid]){
            return FindPivot(arr,mid+1,high);
        }
        return FindPivot(arr,low,mid-1);

    }
    public static int bsearch(int[] arr, int low, int high, int key){
        int mid = (low+high)/2;
        if (low > high){
            System.out.println("OOPS!NOT FOUND");
            return -1;

        }
        else if (arr[mid] == key){
            System.out.println("Element found at " + mid);
            return (mid);
        }
        else if(arr[mid] > key){
            return bsearch(arr,low,mid-1,key);
        }
        return bsearch(arr,mid+1,high,key);
    }
}